File: dotnet/src/Botas/SuggestedActions.cs
using System.Text.Json.Serialization;

namespace Botas;

/// <summary>
/// Represents an actionable button that can be presented to the user.
/// Used within <see cref="SuggestedActions"/> or as part of card content.
/// </summary>
public class CardAction
{
    /// <summary>Action type (e.g. <c>"imBack"</c>, <c>"postBack"</c>, <c>"openUrl"</c>, <c>"messageBack"</c>). Defaults to <c>"imBack"</c>.</summary>
    [JsonPropertyName("type")] public string Type { get; set; } = "imBack";

    /// <summary>Text displayed on the button.</summary>
    [JsonPropertyName("title")] public string? Title { get; set; }

    /// <summary>Value sent to the bot when the action is triggered.</summary>
    [JsonPropertyName("value")] public string? Value { get; set; }

    /// <summary>Text sent to the bot as a message when the action is triggered (for <c>"imBack"</c> and <c>"postBack"</c>).</summary>
    [JsonPropertyName("text")] public string? Text { get; set; }

    /// <summary>Text displayed in the chat feed after the action is triggered.</summary>
    [JsonPropertyName("displayText")] public string? DisplayText { get; set; }

    /// <summary>URL of an image to display on the button.</summary>
    [JsonPropertyName("image")] public string? Image { get; set; }

    /// <summary>Extension data dictionary that preserves unknown JSON properties during round-trip serialization.</summary>
    [JsonExtensionData] public Dictionary<string, object?> Properties { get; set; } = [];
}

/// <summary>
/// A set of suggested actions that the channel can present to the user as quick-reply buttons.
/// </summary>
public class SuggestedActions
{
    /// <summary>Array of user IDs that should receive the suggested actions. When <c>null</c>, all participants see them.</summary>
    [JsonPropertyName("to")] public string[]? To { get; set; }

    /// <summary>The card actions to present as suggested responses.</summary>
    [JsonPropertyName("actions")] public CardAction[] Actions { get; set; } = [];

    /// <summary>Extension data dictionary that preserves unknown JSON properties during round-trip serialization.</summary>
    [JsonExtensionData] public Dictionary<string, object?> Properties { get; set; } = [];
}

File: node/packages/botas/src/suggested-actions.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * A quick-reply button or action displayed to the user.
 *
 * Card actions represent clickable options in clients that support suggested actions
 * or hero/thumbnail cards.
 */
export interface CardAction {
  /** Action type (e.g. `"imBack"`, `"postBack"`, `"openUrl"`, `"messageBack"`). */
  type: string
  /** Display text on the button. */
  title?: string
  /** Value sent back to the bot when the action is invoked. */
  value?: string
  /** Text to insert into the compose box when the action is invoked. */
  text?: string
  /** Text shown in the chat feed when the action is invoked (may differ from `text`). */
  displayText?: string
  /** URL to an icon image displayed on the button. */
  image?: string
  /** Additional channel-specific properties. */
  [key: string]: unknown
}

/**
 * Suggested actions (quick-reply buttons) attached to an activity.
 *
 * When present, the channel renders the actions as tappable buttons below
 * the message. Actions are typically removed after the user selects one.
 */
export interface SuggestedActions {
  /** Recipient IDs that should see the suggested actions. Empty or omitted means all participants. */
  to?: string[]
  /** The list of actions to display. */
  actions: CardAction[]
  /** Additional channel-specific properties. */
  [key: string]: unknown
}
